app/forms/declaration/DepartureTransport.scala

```diff
@@ -33,60 +32,24 @@ object DepartureTransport extends DeclarationPage {
-
-  private def meansOfTransportOnDepartureIDNumberMapping(transportType: String) =
-    s"${meansOfTransportOnDepartureIDNumberKey}_$transportType" -> mandatoryIfEqual(
-      meansOfTransportOnDepartureTypeKey,
-      transportType,
-      text()
-        .verifying("declaration.transportInformation.meansOfTransport.reference.error.empty", nonEmpty)
-        .verifying("declaration.transportInformation.meansOfTransport.reference.error.length", isEmpty or noLongerThan(35))
-        .verifying(
-          "declaration.transportInformation.meansOfTransport.reference.error.invalid",
-          isEmpty or (noLongerThan(35) and isAlphanumericWithAllowedSpecialCharacters)
-        )
-    )
+  val prefix = "declaration.transportInformation.meansOfTransport.departure"
+
+  def form(transportCodes: Seq[TransportCode]): Form[DepartureTransport] = Form(mappingFor(transportCodes))
 
-  private val requiredMapping = mapping(
-    meansOfTransportOnDepartureTypeKey -> meansOfTransportOnDepartureTypeRequired,
-    meansOfTransportOnDepartureIDNumberMapping(IMOShipIDNumber),
-    meansOfTransportOnDepartureIDNumberMapping(NameOfVessel),
-    meansOfTransportOnDepartureIDNumberMapping(WagonNumber),
-    meansOfTransportOnDepartureIDNumberMapping(VehicleRegistrationNumber),
-    meansOfTransportOnDepartureIDNumberMapping(IATAFlightNumber),
-    meansOfTransportOnDepartureIDNumberMapping(AircraftRegistrationNumber),
-    meansOfTransportOnDepartureIDNumberMapping(EuropeanVesselIDNumber),
-    meansOfTransportOnDepartureIDNumberMapping(NameOfInlandWaterwayVessel)
-  )(form2Model)(model2Form)
-
-  private val optionalMapping = mapping(
-    meansOfTransportOnDepartureTypeKey -> meansOfTransportOnDepartureTypeOptional,
-    meansOfTransportOnDepartureIDNumberMapping(IMOShipIDNumber),
-    meansOfTransportOnDepartureIDNumberMapping(NameOfVessel),
-    meansOfTransportOnDepartureIDNumberMapping(WagonNumber),
-    meansOfTransportOnDepartureIDNumberMapping(VehicleRegistrationNumber),
-    meansOfTransportOnDepartureIDNumberMapping(IATAFlightNumber),
-    meansOfTransportOnDepartureIDNumberMapping(AircraftRegistrationNumber),
-    meansOfTransportOnDepartureIDNumberMapping(EuropeanVesselIDNumber),
-    meansOfTransportOnDepartureIDNumberMapping(NameOfInlandWaterwayVessel)
-  )(form2Model)(model2Form)
+  private def mappingFor(transportCodes: Seq[TransportCode]): Mapping[DepartureTransport] =
+    mapping(
+      radioButtonGroupId -> mappingForRadioDepartureType(transportCodes),
+      mappingForInputDepartureNumber(transportCodes(0)),
+      mappingForInputDepartureNumber(transportCodes(1)),
+      mappingForInputDepartureNumber(transportCodes(2)),
+      mappingForInputDepartureNumber(transportCodes(3)),
+      mappingForInputDepartureNumber(transportCodes(4)),
+      mappingForInputDepartureNumber(transportCodes(5)),
+      mappingForInputDepartureNumber(transportCodes(6)),
+      mappingForInputDepartureNumber(transportCodes(7))
```

Comment:
These 'magic numbers' risk throwing a `java.lang.IndexOutOfBoundsException` if anyone ever modifies the `transportCodesForV1`etc TransportCodes Lists. Is there a way to avoid this by mapping the Seq values that are sent? :thinking:

Reply:
How does this work with `transportCodesForV3WhenPC0019` that has 9 elements?

Reply:
Sadly form mapping in Play works by passing a specific number of parameters, and not, for instance, passing a vararg parameter. So, no :frowning_face: there is now way to avoid this, or at least it was me not able to find a DRY alternative. The only other possible alternative is to have a mapping for every version of the page (and we have 4 of them).

Reply:
I was wondering if you could use a Tuple8 instead for the 'transportCodesForVX' vals, then if someone changed the number of elements we would get a compiler error. What do you think?

Reply:
This method is for mapping the input fields associated with the radio buttons (8 input fields). In `transportCodesForV3WhenPC0019` the 9th radio has no input field associated.

Reply:
I also thought to see if a tuple could be used but this is not feasible due to the fact that the elements the form receives as parameters could be 8 or 9, and we cannot express that with a tuple.